src/lib.rs

```diff
@@ -1,289 +1,160 @@
-//! attempt. The same result can be achieved by building a `Retry` object incrementally:
+//! let result = retry(Fixed::from_millis(100).take(1), || {
+//!     match collection.next() {
+//!         Some(n) if n == 3 => Ok("n is 3!"),
+//!         Some(_) => Err("n must be 3!"),
+//!         None => Err("n was never 3!"),
+//!     }
+//! });
 //!
-//! ```
-//! # use retry::Retry;
-//! # struct Client;
-//! # impl Client {
-//! #     fn query_job_status(&self) -> Response {
-//! #         Response {
-//! #             code: 200,
-//! #             body: "success",
-//! #         }
-//! #     }
-//! # }
-//! # struct Response {
-//! #     code: u16,
-//! #     body: &'static str,
-//! # }
-//! # let api_client = Client;
-//! match Retry::new(
-//!     &mut || api_client.query_job_status(),
-//!     &mut |response| response.code == 200
-//! ).try(10).wait(500).execute() {
-//!     Ok(response) => println!("Job completed with result: {}", response.body),
-//!     Err(error) => println!("Job completion could not be verified: {}", error),
-//! }
+//! assert!(result.is_err());
 //! ```
 
+#![deny(missing_debug_implementations)]
+#![deny(missing_docs)]
+
 extern crate rand;
 
-use std::error::Error;
+use std::error::Error as StdError;
 use std::fmt::Error as FmtError;
 use std::fmt::{Display,Formatter};
 use std::thread::sleep;
 use std::time::Duration;
 
-use rand::distributions::{IndependentSample, Range};
-use rand::thread_rng;
-
-/// Builder object for a retryable operation.
-#[derive(Debug)]
-pub struct Retry<'a, F: FnMut() -> R + 'a, G: FnMut(&R) -> bool + 'a, R> {
-    condition_fn: &'a mut G,
-    tries: Option<u64>,
-    value_fn: &'a mut F,
-    timeout: Option<u64>,
-    wait: Wait,
-}
-
-impl<'a, F: FnMut() -> R, G: FnMut(&R) -> bool, R> Retry<'a, F, G, R> {
-    /// Builds a new `Retry` object.
-    pub fn new(
-        value_fn: &'a mut F,
-        condition_fn: &'a mut G
-    ) -> Retry<'a, F, G, R> where F: FnMut() -> R, G: FnMut(&R) -> bool {
-        Retry {
-            condition_fn: condition_fn,
-            tries: None,
-            value_fn: value_fn,
-            wait: Wait::None,
-            timeout: None,
-        }
-    }
-
-    /// Begins executing the retryable operation.
-    pub fn execute(self) -> Result<R, RetryError> {
-        if self.tries.is_some() && self.tries.unwrap() == 0 {
-            return Err(RetryError { message: "tries must be non-zero" });
-        }
-
-        let mut try = 0;
-
-        loop {
-            if self.tries.is_some() && self.tries.unwrap() == try {
-                return Err(RetryError { message: "reached last try without condition match" })
+pub mod delay;
+
+/// Retry the given operation synchronously until it succeeds, or until the given `Duration`
+/// iterator ends.
+pub fn retry<I, O, R, E>(mut iterator: I, mut operation: O) -> Result<R, Error<E>>
+where I: Iterator<Item=Duration>, O: FnMut() -> Result<R, E> {
```

Comment:
It might be useful to go with `IntoIterator` instead of `Iterator` here, to grant more flexibility:

```
pub fn retry<I, O, R, E>(iterable: I, mut operation: O) -> Result<R, Error<E>> where I: IntoIterator<Item=Duration>, O: FnMut() -> Result<R, E> {
  let mut iterator = iterable.into_iter();
  ...
}
```